phases/19-capstone-projects/42-large-corpus-downloader/code/main.py

```
@@ -0,0 +1,514 @@
+                    buf = fh.read(min(self.chunk_bytes, remaining))
+                    if not buf:
+                        break
+                    rolling.update(buf)
+                    remaining -= len(buf)
+        else:
+            if shard_path.exists():
+                shard_path.unlink()
+            if checkpoint_path.exists():
+                checkpoint_path.unlink()
+
+        request = urllib.request.Request(plan.url)
```

Ruff (0.15.14)

[error] Audit URL open for permitted schemes. Allowing use of `file:` or custom schemes is often unexpected.

(S310)